File: examples/demo-server.ts
/**
 * mcp-paginate demo server
 *
 * Shows three realistic tools that return large payloads.
 * Without mcp-paginate these would overflow context windows;
 * with it the LLM pages through seamlessly.
 *
 * Run:
 *   npx tsx examples/demo-server.ts
 *
 * Then connect any MCP client (Claude Desktop, mcp-cli, etc.)
 * and call list_records, list_files, or fetch_logs.
 */

import { McpServer } from "@modelcontextprotocol/sdk/server/mcp.js";
import { StdioServerTransport } from "@modelcontextprotocol/sdk/server/stdio.js";
import { z } from "zod";
import { paginate } from "../src/index.js";

// ─── Server setup ───────────────────────────────────────────────────────────

const server = new McpServer({
  name: "mcp-paginate-demo",
  version: "0.2.0",
});

/**
 * BEFORE mcp-paginate: the LLM receives a truncated response or an error
 * when tool output exceeds its context window.
 *
 * AFTER (one line): every tool below automatically paginates.
 */
paginate(server, {
  maxTokens: 1000, // low for demo — use 4000+ in production
});

// ─── Tool 1: large database query ───────────────────────────────────────────

interface Record {
  id: number;
  name: string;
  email: string;
  department: string;
  salary: number;
  startDate: string;
}

function generateEmployees(count: number): Record[] {
  const departments = ["Engineering", "Sales", "Marketing", "HR", "Finance"];
  return Array.from({ length: count }, (_, i) => ({
    id: i + 1,
    name: `Employee ${i + 1}`,
    email: `employee${i + 1}@company.example`,
    department: departments[i % departments.length]!,
    salary: 60_000 + (i % 50) * 1_000,
    startDate: new Date(Date.now() - i * 30 * 86_400_000).toISOString().slice(0, 10),
  }));
}

server.tool(
  "list_records",
  "List employee records from the HR database",
  {
    limit: z
      .number()
      .int()
      .min(1)
      .max(2000)
      .optional()
      .describe("Number of records to return (default: 500)"),
  },
  async ({ limit = 500 }) => {
    const employees = generateEmployees(limit);
    return {
      content: [{ type: "text" as const, text: JSON.stringify(employees, null, 2) }],
    };
  }
);

// ─── Tool 2: recursive file listing ─────────────────────────────────────────

server.tool(
  "list_files",
  "Recursively list files under a path",
  {
    path: z.string().describe("Directory path to scan"),
    depth: z
      .number()
      .int()
      .min(1)
      .max(5)
      .optional()
      .describe("Max recursion depth (default: 3)"),
  },
  async ({ path, depth = 3 }) => {
    const entries: Array<{ path: string; type: string; sizeBytes: number; modified: string }> = [];
    const fileCount = Math.pow(8, Math.min(depth, 3)); // scales with depth

    for (let i = 0; i < fileCount; i++) {
      const dir = `${path}/subdir-${Math.floor(i / 8)}`;
      entries.push({
        path: `${dir}/file-${i}.ts`,
        type: "file",
        sizeBytes: 1024 + (i * 137) % 102400,
        modified: new Date(Date.now() - i * 3_600_000).toISOString(),
      });
    }

    return {
      content: [{ type: "text" as const, text: JSON.stringify(entries, null, 2) }],
    };
  }
);

// ─── Tool 3: application log fetch ──────────────────────────────────────────

const LOG_LEVELS = ["INFO", "WARN", "ERROR", "DEBUG"] as const;

server.tool(
  "fetch_logs",
  "Fetch recent application logs for a service",
  {
    service: z.string().describe("Service name (e.g. api-gateway, auth-service)"),
    lines: z
      .number()
      .int()
      .min(1)
      .max(5000)
      .optional()
      .describe("Number of log lines (default: 500)"),
    level: z
      .enum(["INFO", "WARN", "ERROR", "DEBUG", "ALL"])
      .optional()
      .describe("Filter by log level (default: ALL)"),
  },
  async ({ service, lines = 500, level = "ALL" }) => {
    const logs = Array.from({ length: lines }, (_, i) => {
      const logLevel = LOG_LEVELS[i % LOG_LEVELS.length]!;
      if (level !== "ALL" && logLevel !== level) return null;
      return {
        timestamp: new Date(Date.now() - i * 1_000).toISOString(),
        level: logLevel,
        service,
        traceId: Math.random().toString(36).slice(2, 18),
        message: `[${service}] Request ${i + 1} processed in ${10 + (i % 200)}ms — status ${i % 20 === 0 ? 500 : 200}`,
      };
    }).filter(Boolean);

    return {
      content: [{ type: "text" as const, text: JSON.stringify(logs, null, 2) }],
    };
  }
);

// ─── Connect ─────────────────────────────────────────────────────────────────

const transport = new StdioServerTransport();
await server.connect(transport);
console.error(
  "[demo-server] Running. Tools: list_records | list_files | fetch_logs | get_next_page"
);

File: python/examples/demo_server.py
"""
mcp-pager Python demo server.

Run:
    python examples/demo_server.py

Then connect any MCP client and call list_records, list_files, or fetch_logs.
Each tool returns large payloads that mcp-pager automatically pages.
"""
import asyncio
import json
import random
import string
from datetime import datetime, timedelta

from mcp.server.fastmcp import FastMCP

from mcp_pager import paginate

mcp = FastMCP("mcp-pager-demo")

# One line — every tool below is now token-safe.
paginate(mcp, max_tokens=4000, ttl_ms=10 * 60 * 1000)


@mcp.tool(description="List employee records from the HR database")
async def list_records(limit: int = 500) -> str:
    departments = ["Engineering", "Sales", "Marketing", "HR", "Finance"]
    records = [
        {
            "id": i + 1,
            "name": f"Employee {i + 1}",
            "email": f"employee{i + 1}@company.example",
            "department": departments[i % len(departments)],
            "salary": 60_000 + (i % 50) * 1_000,
            "start_date": (datetime.now() - timedelta(days=i * 30)).date().isoformat(),
        }
        for i in range(min(limit, 2000))
    ]
    return json.dumps(records, indent=2)


@mcp.tool(description="Recursively list files under a path")
async def list_files(path: str, depth: int = 3) -> str:
    count = min(8 ** depth, 512)
    entries = [
        {
            "path": f"{path}/subdir-{i // 8}/file-{i}.py",
            "type": "file",
            "size_bytes": 1024 + (i * 137) % 102400,
            "modified": (datetime.now() - timedelta(hours=i)).isoformat(),
        }
        for i in range(count)
    ]
    return json.dumps(entries, indent=2)


@mcp.tool(description="Fetch recent application logs for a service")
async def fetch_logs(service: str, lines: int = 500, level: str = "ALL") -> str:
    levels = ["INFO", "WARN", "ERROR", "DEBUG"]
    logs = [
        {
            "timestamp": (datetime.now() - timedelta(seconds=i)).isoformat(),
            "level": levels[i % 4],
            "service": service,
            "trace_id": "".join(random.choices(string.hexdigits, k=16)),
            "message": f"[{service}] Request {i + 1} processed in {10 + i % 200}ms — status {500 if i % 20 == 0 else 200}",
        }
        for i in range(min(lines, 5000))
        if level == "ALL" or levels[i % 4] == level
    ]
    return json.dumps(logs, indent=2)


if __name__ == "__main__":
    mcp.run()
